cpp/src/binaryop/compiled/struct_binary_ops.cuh

```diff
@@ -0,0 +1,136 @@
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include "binary_ops.hpp"
+#include "operation.cuh"
+
+#include <cudf/binaryop.hpp>
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/column/column_view.hpp>
+#include <cudf/detail/iterator.cuh>
+#include <cudf/table/experimental/row_operators.cuh>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/exec_policy.hpp>
+
+namespace cudf::binops::compiled::detail {
+template <class T, class... Ts>
+inline constexpr bool is_any_v = std::disjunction<std::is_same<T, Ts>...>::value;
+
+template <typename DeviceComparator, class AdaptableUnaryFunction, class Iterator>
+struct device_comparison_functor {
+  using optional_iterator = thrust::transform_iterator<AdaptableUnaryFunction, Iterator>;
```

Comment:
This name alias is misleading. It is actually not an optional iterator, isn't it? Why not just `transform_iterator` or some other name?

Reply:
It is, the optional iterator is a specialization of the transform_iterator. It's a few layers deep because it's hidden under a few layers of `auto` types and optional iterators aren't used in this context much.

Either way, the cleaner approach is to template the iterator rather than the nested template types.
```
template <typename DeviceComparator, typename OptionalIterator>
```